torch/utils/data/sampler.py

```diff
@@ -210,3 +210,76 @@ def __len__(self):
             return len(self.sampler) // self.batch_size
         else:
             return (len(self.sampler) + self.batch_size - 1) // self.batch_size
+
+class DistributedChunkSampler(Sampler):
```

Comment:
I pushed a revision with both samplers (modified DistributedSampler and a new DistributedChunkSampler) and we discussed in person the DistributedSampler issue, which was having all constructors arguments to be optional. For that reason, the user can instantiate the sampler in an invalid state right from beginning. Because objects should always be instantiated in a valid state, you agreed to drop  that approach :). You can still see the code changes for DistributedSampler at 7f26909 and we can discuss further
Additionally to the constructor issue, the source code would almost double just to add lots of if/else to prevent indices padding and to use the new parameter num_indices instead of the dataset one. Documentation would have to be carefully rewritten to explain the duality of the sampler too.